### ISSUE TITLE
Padrino (v. 0.15.3) fails to start with Ruby 3.3.0: uninitialized constant Padrino::Rackup (NameError)

### ISSUE THREAD

**Issue by kimberlykeilymh** [padrino/padrino-framework#2288]
#### Do you want to request a *feature* or report a *bug*?

Bug

#### What is the current behavior?

Failure to start server using Ruby 3.3.0. Below is the stack trace.

```
$ bundle exec padrino s
bundler: failed to load command: padrino (/home/kimberly/.gem/ruby/3.3.0/bin/padrino)
/home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-d75858b05d9c/padrino-core/lib/padrino-core/server.rb:33:in `<module:Padrino>': uninitialized constant Padrino::Rackup (NameError)

  class Server < Rackup::Server
                 ^^^^^^
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-d75858b05d9c/padrino-core/lib/padrino-core/server.rb:1:in `<top (required)>'
	from /home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `require'
	from /home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `block (2 levels) in replace_require'
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-d75858b05d9c/padrino-core/lib/padrino-core.rb:15:in `<top (required)>'
	from /home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `require'
	from /home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `block (2 levels) in replace_require'
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-d75858b05d9c/padrino/lib/padrino.rb:3:in `<top (required)>'
	from <internal:/home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
	from <internal:/home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
	from /home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `block (2 levels) in replace_require'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/runtime.rb:60:in `block (2 levels) in require'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/runtime.rb:55:in `each'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/runtime.rb:55:in `block in require'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/runtime.rb:44:in `each'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/runtime.rb:44:in `require'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler.rb:187:in `require'
	from /home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `require'
	from /home/kimberly/.rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `block (2 levels) in replace_require'
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-d75858b05d9c/padrino-core/lib/padrino-core/cli/launcher.rb:25:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/base.rb:584:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-d75858b05d9c/padrino-core/bin/padrino:9:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/bin/padrino:25:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/bin/padrino:25:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli/exec.rb:58:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli/exec.rb:58:in `kernel_load'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli/exec.rb:23:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli.rb:451:in `exec'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli.rb:34:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli.rb:28:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/exe/bundle:28:in `block in <top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/exe/bundle:20:in `<top (required)>'
	from /home/kimberly/.rubies/ruby-3.3.0/bin/bundle:25:in `load'
	from /home/kimberly/.rubies/ruby-3.3.0/bin/bundle:25:in `<main>'
```

#### What is the expected behavior?

Start the server without errors.

#### Which versions of Ruby, Padrino, Sinatra, Rack, OS are you using? Did this work in previous versions?

Ruby 3.3.0, Padrino 0.15.3, Ubuntu 23.10.

**Comment by nesquena** [comment]
@achiurizo would this be fixed with a release from master? Would be great to cut a release that supports Ruby 3.3

**Comment by kimberlykeilymh** [comment]
Something important to mention is that I am using the git repository as a source code, I mean, in my Gemfle it is as follows: `gem 'padrino', git: 'https://github.com/padrino/padrino-framework'`

**Comment by achiurizo** [comment]
@kimberlykeilymh thank you for the report! I've pushed out #2289 which should resolve this (adds `rackup`). If you want to test this sooner, you can try using the branch directly until its merged:

```
gem 'padrino', git: 'https://github.com/padrino/padrino-framework', branch: 'achiurizo/add-rackup-to-gemspec'
```

@nesquena we should do more testing after #2289 but after that I think we can go ahead and cut a release

**Comment by nesquena** [comment]
I went ahead and merged that change. Let us know if that fixes things for you. If everything is working as expected, we can do final testing and cut a release

**Comment by kimberlykeilymh** [comment]
Thanks guys! The change fixed the bug with `rackup`. 

Although now I get this other error:

```
bundler: failed to load command: padrino (/home/kimberly/.gem/ruby/3.3.0/bin/padrino)
/home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-889f6623c071/padrino-core/lib/padrino-core/server.rb:24:in `detect_application': undefined method `merge' for nil (NoMethodError)

      [rack_app, rack_options.merge(options)]
                             ^^^^^^
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-889f6623c071/padrino-core/lib/padrino-core/server.rb:12:in `run!'
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-889f6623c071/padrino-core/lib/padrino-core/cli/adapter.rb:9:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-889f6623c071/padrino-core/lib/padrino-core/cli/launcher.rb:30:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/base.rb:584:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/bundler/gems/padrino-framework-889f6623c071/padrino-core/bin/padrino:9:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/bin/padrino:25:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/bin/padrino:25:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli/exec.rb:58:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli/exec.rb:58:in `kernel_load'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli/exec.rb:23:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli.rb:451:in `exec'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli.rb:34:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/cli.rb:28:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/exe/bundle:28:in `block in <top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.9/exe/bundle:20:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/bin/bundle:25:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/bin/bundle:25:in `<main>'

```

Is it also possible to check this error? I can open another issue and close this one if you want.

**Comment by nesquena** [comment]
@achiurizo hmm, looks like rack_options is nil rather than a hash now in the server start sequence

**Comment by srolandmarshall** [comment]
@nesquena I built a Padrino app from scratch with limited changes to config (updating database) — using Edge release w/ the inclusion of Backup in the Gemspec results in the issue @kimberlykeilymh mentioned above.

```
➜  probable-pitchers-padrino git:(main) padrino start
/Users/sam-home/.rvm/gems/ruby-3.1.3/bundler/gems/padrino-framework-889f6623c071/padrino-core/lib/padrino-core/server.rb:24:in `detect_application': undefined method `merge' for nil:NilClass (NoMethodError)

      [rack_app, rack_options.merge(options)]
                             ^^^^^^
        from /Users/sam-home/.rvm/gems/ruby-3.1.3/bundler/gems/padrino-framework-889f6623c071/padrino-core/lib/padrino-core/server.rb:12:in `run!'
```

**Comment by nesquena** [comment]
As a starting point, thanks @srolandmarshall I don't want to be too optimistic but it seems like part of this may just be a change to the way Rack parses the files. I had rack options default to {} on https://github.com/padrino/padrino-framework/pull/2290 (branch [nesquena-rack-options-fix](https://github.com/padrino/padrino-framework/tree/nesquena-rack-options-fix)) until achiu may have a chance to test end to end

**Comment by srolandmarshall** [comment]
Checked it out, seemed to solve the problem on my end. Thanks for the quick patch @nesquena

**Comment by nesquena** [comment]
Great, thanks for testing!

**Comment by kimberlykeilymh** [comment]
Hi guys! Yes, that solved the problem and padrino is running without errors. Thanks very much for the fix!

**Comment by nesquena** [comment]
That's great news! Since two of you confirmed, perhaps it's time to cut a new release soon

**Comment by floriandejonckheere** [comment]
Hi! Any news on a new release of Padrino?

**Comment by nesquena** [comment]
I'll push out a new release with these fixes within a week

**Comment by nesquena** [comment]
Okay I pushed out [0.16.0.pre2](https://rubygems.org/gems/padrino-core/versions/0.16.0.pre2) - can one or two of you try this and report back confirming everything works as it should?

@floriandejonckheere @achiurizo @kimberlykeilymh @srolandmarshall 

As soon as I get confirmations, I'll cut it as an official release! If anyone wants to help fix the CI to pass the tests fully, that would be appreciated as well 😅

**Comment by kimberlykeilymh** [comment]
Hi folks! Thanks for pushed out a new version!
I try the pre-release version with Ruby 3.3.0, but I got this error:

`
bundler: failed to load command: padrino (/home/kimberly/.gem/ruby/3.3.0/bin/padrino)
/home/kimberly/.gem/ruby/3.3.0/gems/padrino-core-0.16.0.pre2/lib/padrino-core/server.rb:86:in `detect_rack_handler': Server handler (thin, puma, spider-gazelle, mongrel, trinidad, webrick) not found. (RuntimeError)
	from /home/kimberly/.gem/ruby/3.3.0/gems/padrino-core-0.16.0.pre2/lib/padrino-core/server.rb:45:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/padrino-core-0.16.0.pre2/lib/padrino-core/server.rb:12:in `run!'
	from /home/kimberly/.gem/ruby/3.3.0/gems/padrino-core-0.16.0.pre2/lib/padrino-core/cli/adapter.rb:9:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/padrino-core-0.16.0.pre2/lib/padrino-core/cli/launcher.rb:30:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/thor-1.3.1/lib/thor/base.rb:584:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/padrino-core-0.16.0.pre2/bin/padrino:9:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/bin/padrino:25:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/bin/padrino:25:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/cli/exec.rb:58:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/cli/exec.rb:58:in `kernel_load'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/cli/exec.rb:23:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/cli.rb:455:in `exec'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/cli.rb:35:in `dispatch'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/cli.rb:29:in `start'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/exe/bundle:28:in `block in <top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
	from /home/kimberly/.gem/ruby/3.3.0/gems/bundler-2.5.17/exe/bundle:20:in `<top (required)>'
	from /home/kimberly/.gem/ruby/3.3.0/bin/bundle:25:in `load'
	from /home/kimberly/.gem/ruby/3.3.0/bin/bundle:25:in `<main>'
`

**Comment by kimberlykeilymh** [comment]
Something I forgot to mention is that we use WEBrick as a server, which is part of the server list.

**Comment by kimberlykeilymh** [comment]
Trying with Puma and thin I got the same error

**Comment by nesquena** [comment]
Thanks @kimberlykeilymh for testing!

**Comment by nesquena** [comment]
@achiurizo Any idea what might be going on above?

**Comment by osbre** [comment]
I have found the cause and sent a fix: https://github.com/padrino/padrino-framework/pull/2294

**Comment by nesquena** [comment]
> @nesquena I have found the cause and sent a fix: https://github.com/padrino/padrino-framework/pull/2294
> 
> 
> 
> Can confirm it now works with Puma.

Thank you! 🙏 I'll release pre3 today with this fix so others can confirm

**Comment by nesquena** [comment]
Pushed [pre3](https://rubygems.org/gems/padrino-core/versions/0.16.0.pre3)

@kimberlykeilymh Can you try again with the latest pre3 release and confirm that everything works now?

**Comment by kimberlykeilymh** [comment]
Hi Nathan!
I tested the pre3 version and the error with the server no longer occurs, thank you very much for fixing it!

On the other hand, I had an error with this new version regarding the parameters received by the controller. In one of the forms we have fields nested in two levels that reach the controller in the following way:

```
(ruby) params
{
     "account"=>
     {
          "status"=>"active",
          "account_settings"=>
          {
               "example_field"=>"example_value"
          },
          "name"=>"Account Name"
     } 
}
```

But now those same parameters are obtained like this:

```
(ruby) params
{
     "account"=>
     {
          "status"=>"active",
          "account_settings[example_field"=>
          {
               "]"=>"example_value"
          },
          "name"=>"Account Name"
     } 
}
```

At the form level the name of the fields remains the same (`account[account_settings[example_field]]`).

This would be for example for an `Account` model, which is associated with an `AccountSettings` model which has an `example_field` field.

Do you have any idea what could be happening? Is there some new definition in the framework that does not allow nesting in more than one level?

Thanks in advance,

**Comment by nesquena** [comment]
Thanks so much for testing @kimberlykeilymh and catching the issue with nested form parameter processing. That seems like a bug. Before we release, we'll fix this so that it matches the original behavior

If anyone can help investigate, and or write tests for this, or fix any other failing tests, that would be appreciated 

Before this next major release, we'll also need to fix all failing tests and test a number of apps work as expected

**Comment by rasheqrahman** [comment]
Building on @kimberlykeilymh's issue, here's what I'm getting when I try to deploy to heroku running Ruby 3.2.5 

NameError: uninitialized constant Rack::Server (NameError)

  class Server < Rack::Server
                     ^^^^^^^^
Did you mean?  TCPServer
/Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/lib/padrino-core/server.rb:34:in `<module:Padrino>'
/Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/lib/padrino-core/server.rb:1:in `<top (required)>'
<internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
<internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
/Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/lib/padrino-core.rb:15:in `<top (required)>'
<internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
<internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
/Users/rasheqrahman/wikiidea/Rakefile:4:in `<top (required)>'
/Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/rake-13.2.1/exe/rake:27:in `<top (required)>'
/Users/rasheqrahman/.rvm/gems/ruby-3.2.5/bin/ruby_executable_hooks:22:in `eval'
/Users/rasheqrahman/.rvm/gems/ruby-3.2.5/bin/ruby_executable_hooks:22:in `<main>'
(See full trace by running task with --trace)

**Comment by rasheqrahman** [comment]
Can anyone help? I and ChatGPT are at a loss on how to fix this. I've just been trying to update my sinatra/padrino app to run on the latest heroku stack which runs 3.2.5

**Comment by rasheqrahman** [comment]
I definitely keep getting the same error and I'm stuck.


/Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/lib/padrino-core/server.rb:34:in `<module:Padrino>': uninitialized constant Rack::Server (NameError)

  class Server < Rack::Server
                     ^^^^^^^^
Did you mean?  TCPServer
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/lib/padrino-core/server.rb:1:in `<top (required)>'
	from <internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from <internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/lib/padrino-core.rb:15:in `<top (required)>'
	from <internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from <internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-0.15.3/lib/padrino.rb:3:in `<top (required)>'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/bundler-2.5.21/lib/bundler/runtime.rb:60:in `require'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/bundler-2.5.21/lib/bundler/runtime.rb:60:in `block (2 levels) in require'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/bundler-2.5.21/lib/bundler/runtime.rb:55:in `each'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/bundler-2.5.21/lib/bundler/runtime.rb:55:in `block in require'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/bundler-2.5.21/lib/bundler/runtime.rb:44:in `each'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/bundler-2.5.21/lib/bundler/runtime.rb:44:in `require'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/bundler-2.5.21/lib/bundler.rb:212:in `require'
	from /Users/rasheqrahman/wikiidea/config/boot.rb:14:in `<top (required)>'
	from <internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from <internal:/Users/rasheqrahman/.rvm/rubies/ruby-3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/lib/padrino-core/cli/launcher.rb:25:in `start'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/thor-1.3.2/lib/thor/command.rb:28:in `run'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/thor-1.3.2/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/thor-1.3.2/lib/thor.rb:538:in `dispatch'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/thor-1.3.2/lib/thor/base.rb:584:in `start'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/gems/padrino-core-0.15.3/bin/padrino:9:in `<top (required)>'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/bin/padrino:25:in `load'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/bin/padrino:25:in `<main>'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/bin/ruby_executable_hooks:22:in `eval'
	from /Users/rasheqrahman/.rvm/gems/ruby-3.2.5/bin/ruby_executable_hooks:22:in `<main>'

**Comment by chaymaeBZ** [comment]
@rasheqrahman you are using padrino 0.15.3 and as discussed here, 0.16.0.pre3 should fix `uninitialized constant Rack::Server (NameError)` and introduces support to ruby 3.2.

**Comment by chaymaeBZ** [comment]
Hi @kimberlykeilymh :wave: I am not able to replicate the error you are seeing with nested models. I created a dummy app with Padrino 0.16.0.pre3 version with nested models and forms. However, I am receiving the nested params in the correct way. Could it be something with the form helpers you are using? I used `form_for` for the parent model and `fields_for` for the nested model. Could you share a snippet of your forms?

**Comment by nesquena** [comment]
Thanks for testing this @chaymaeBZ, I'll be happy to release the next version in full outside of pre once we feel confident tests are passing again and that all critical bugs have been worked out. I haven't had time to dig into the nested forms issue yet, I wasn't able to reproduce with a simple nested example, but perhaps it's like second layer nesting where it's failing?

**Comment by chaymaeBZ** [comment]
Thanks @nesquena! indeed I could replicate this!

I wrote simple steps to reproduce it :

`$ padrino g project nested_params_example`
`$ echo "gem 'puma' >> Gemfile"`
( Also Make sure padrino version is 0.16.pre3 in the Gemfile)
`$ bundle exec padrino g controller example`
Edit the example controller :
```ruby
NestedParamsExample::App.controllers :example do
  post :submit, :csrf_protection => false do
    "params received: #{params.inspect}"
  end
end
```

`$ bundle exec padrino s`
`$ curl -X POST http://127.0.0.1:3000/example/submit -d "account[account_settings[example_field]]=value"`
returns :
` # => params received: {"account"=>{"account_settings[example_field"=>{"]"=>"value"}}}`

**Comment by nesquena** [comment]
Thank you, this may be one of the last major issues before we can release the new version officially. If anyone has a chance to review and put in a fix for this, I'd be happy to review and merge and cut a release once kimberly confirms the fix. I'll try to find time to fix it myself eventually. Also any work to fix broken tests is appreciated too

Thanks to everyone for your help

**Comment by chaymaeBZ** [comment]
I did a deep dive into this. As I suspected, this has nothing to do with Padrino. This behavior comes from Rack. And this change was introduced in Rack 3. You can try the same workflow with Sinatra 4 and it will yield the same 'weird' result.

Now this not a bug. In fact, it's only weird to set nested params as `account[account_settings[example_field]]=value`, even though this was tolerated before. You can even set nested params as `account[account_settings]example_field=value` and it will work just fine.

The recommended way to send your nested fields is to use `account[account_settings][example_field]=value`. In fact, this is what the form builders are doing to build nested fields. it sets the name to `account[account_settings][example_field]=value`.

You can try it, switching between `rack 2` and `rack 3`

```ruby
# rack 3
Rack::Utils.parse_nested_query "account[account_settings[example_field]]=value"
#=> {"account"=>{"account_settings[example_field"=>{"]"=>"value"}}}

# rack 2
Rack::Utils.parse_nested_query "account[account_settings[example_field]]=value"
#=> {"account"=>{"account_settings"=>{"example_field"=>"value"}}}

Rack::Utils.parse_nested_query "account[account_settings]example_field=value"
#=> {"account"=>{"account_settings"=>{"example_field"=>"value"}}}
```

The change has been introduced [here](https://github.com/rack/rack/commit/c1e5fbbb59101c039e8b657c8052e152c572d5ac#diff-69c5b6856deb050ba393562673e0cc863d520360df07503081064f531e852bc2R110), where it tries to "scan" for what's between `[` and `]` to set the nested key.

Here is another way to test this behavior of rack

```ruby
require 'stringio'
require 'rack'


def get_params(input)
	rack_input = StringIO.new(input)

	env = {'rack.input' => rack_input, 'REQUEST_METHOD' => 'POST'}
	request = Rack::Request.new(env)

	request.params
end

correct_input = "account[account_settings][example_field]=value"
wrong_input = "account[account_settings[example_field]]=value"

p get_params(correct_input)
# => Rack 2 :
# => {"account"=>{"account_settings"=>{"example_field"=>"value"}}}

# => Rack 3 :
# => {"account"=>{"account_settings"=>{"example_field"=>"value"}}}

p get_params(wrong_input)
# => Rack 2 :
# => {"account"=>{"account_settings"=>{"example_field"=>"value"}}}

# => Rack 3 :
# => {"account"=>{"account_settings[example_field"=>{"]"=>"value"}}}
```

**Comment by nesquena** [comment]
I see, is it possible then that this is more of a won't fix. It may not be our responsibility to address this unless our own form helpers are generating incorrect code. What's your recommendation?

AFAIK this may be the last known issue before I can cut a release.

**Comment by chaymaeBZ** [comment]
I'd say it's a won't fix thing. I wouldn't even call it a bug 😂 `A[B][C]` is how you do nesting, not `A[B[C]]`.

I tried with the form generators, it's not generating such weird nesting syntax. So I'd say it's up to people to fix any weird nesting syntax they hardcoded.

**Comment by chaymaeBZ** [comment]
@kimberlykeilymh can you share a snippet of your form?

**Comment by kimberlykeilymh** [comment]
Hello everyone! Thank you for taking the time to review the potential error.
I’m sharing a snippet of the original form and a modified one for testing (along with their results):

```ruby
= form_for @account, url_for(:accounts, :index, id: @account.id), method: :put do |f|
      .form-group.row
        = f.label 'account_settings[ex_integration_enabled]', caption: 'Integración EX', class: ['col-form-label', 'col-lg-2']
        = f.check_box 'account_settings[ex_integration_enabled]', class: ['form-control', 'custom-control-input'], checked: !@account.account_settings&.value&.[]('ex_integration_enabled').nil?

(ruby) params[:account]
{"account_settings[mp_integration_enabled"=>{"]"=>"1"}}
```

```ruby
= form_for @account, url_for(:accounts, :index, id: @account.id), method: :put do |f|
      .form-group.row
        = f.label 'account_settings[ex_integration_enabled]', caption: 'Integración EX', class: ['col-form-label', 'col-lg-2']
            = f.check_box '[account_settings][ex_integration_enabled]', class: ['form-control', 'custom-control-input'], checked: !@account.account_settings&.value&.[]('ex_integration_enabled').nil?
            
(ruby) params[:account]
{"[account_settings"=>{"mp_integration_enabled"=>{"]"=>"1"}}}
```

**Comment by chaymaeBZ** [comment]
@kimberlykeilymh you are using the form builders in a wrong way. If you inspect both forms, the first checkbox's name will look something like `account[account_settings[ex_integration_enabled]]` and the second will look like `account[[account_settings][ex_integration_enabled]]`. As I explained earlier, this is wrong nesting syntax. please use `fields_for ` to build your nested  forms. Or read the docs for alternatives https://padrinorb.com/guides/application-helpers/nested-object-form-support/

**Comment by nesquena** [comment]
Thanks @chaymaeBZ for investigating this! @kimberlykeilymh can you try to make those adjustments to the form helpers and confirm everything works as expected, test around and give me confirmation that nothing appears broken?
